crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2466,16 +2480,138 @@ impl<'db> TypeInferenceBuilder<'db> {
         let ast::ExprCompare {
             range: _,
             left,
-            ops: _,
+            ops,
             comparators,
         } = compare;
 
-        self.infer_expression(left);
-        // TODO actually handle ops and return correct type
-        for right in comparators.as_ref() {
-            self.infer_expression(right);
+        // https://docs.python.org/3/reference/expressions.html#comparisons
+        // > Formally, if `a, b, c, …, y, z` are expressions and `op1, op2, …, opN` are comparison
+        // > operators, then `a op1 b op2 c ... y opN z` is equivalent to a `op1 b and b op2 c and
+        // ... > y opN z`, except that each expression is evaluated at most once.
+        //
+        // As some operators (==, !=, <, <=, >, >=) *can* return an arbitrary type, the logic below
+        // is shared with the one in `infer_binary_type_comparison`.
+        Self::infer_chained_boolean_types(
+            self.db,
+            ast::BoolOp::And,
+            std::iter::once(left.as_ref())
+                .chain(comparators.as_ref().iter())
+                // Evaluate expressions before iterating through pairs with `windows`
+                .map(|expr| self.infer_expression(expr))
+                .collect::<Vec<_>>()
+                .windows(2)
+                //.tuple_windows(2)
```

Comment:
Using `.tuple_windows::<(Type<'db>, Type<'db>)>()` creates some problems for me
- Above, `line 2500`, we borrow `*self` as mutable to perform the `infer_expression`
- Then below, `line 2507`, we borrow again `self` (at the minimum we need to borrow `self.db`)

I couldn't do better than that. I could work around some other borrows (remove `self` from some functions that don't need it, store diagnostics in a `Vec` to emit later, ...) but I'm not sure how to get out of the borrow issue explained above.

As always, suggestions welcome!

Reply:
Are the inferred types cached somehow? Or are we doing the work of inference twice?

Reply:
> Are the inferred types cached somehow?

Yup! In fact, there are various places where the code will panic if it detects that it's tried to infer the type of the same expression twice.

`self.infer_expression()` does this to store the type of the expression in the cache:

https://github.com/astral-sh/ruff/blob/4aefe523938f7176be0bcaa03b6f126c8ae783fb/crates/red_knot_python_semantic/src/types/infer.rs#L1694-L1698

And then `self.types.expression_ty()` would simply look things up in the cache again:

https://github.com/astral-sh/ruff/blob/4aefe523938f7176be0bcaa03b6f126c8ae783fb/crates/red_knot_python_semantic/src/types/infer.rs#L190-L192

Reply:
Amazing, this solved my problems, thank you!!